tc/core/polyhedral/memory_promotion.cc

```diff
@@ -158,25 +188,61 @@ isl::set TensorReferenceGroup::approximateFootprint() const {
   for (size_t i = 0; i < approximation.size(); ++i) {
     auto dim = approximation.at(i);
     auto rhs = isl::aff(lspace, isl::dim_type::set, i);
-    isl::map partial = (isl::aff_map(dim.lowerBound) <= rhs) &
-        (isl::aff_map(dim.lowerBound + dim.size) > rhs);
+    auto lowerBound = dim.lowerBound * dim.stride + dim.offset;
+    auto upperBound = (dim.lowerBound + dim.size) * dim.stride + dim.offset;
+    isl::map partial =
+        (isl::aff_map(lowerBound) <= rhs) & (isl::aff_map(upperBound) > rhs);
+
     accessed = accessed & partial;
   }
   return accessed.range();
 }
 
+namespace {
+template <typename Func>
+isl::multi_aff createMA(
+    const std::vector<ScopedFootprintDim>& dims,
+    Func accessor) {
+  if (dims.size() == 0) {
+    throw promotion::PromotionNYI("promotion for scalars");
+  }
+  auto space =
+      add_range(accessor(dims.at(0)).get_space().domain(), dims.size());
+  auto ma = isl::multi_aff::zero(space);
+
+  int i = 0;
+  for (const auto& a : dims) {
+    ma = ma.set_aff(i++, accessor(a));
+  }
+  return ma;
+}
+} // namespace
+
 isl::multi_aff ScopedFootprint::lowerBounds() const {
+  return createMA(
+      *this, [](const ScopedFootprintDim& dim) { return dim.lowerBound; });
+}
+
+isl::multi_aff ScopedFootprint::offsets() const {
+  return createMA(
+      *this, [](const ScopedFootprintDim& dim) { return dim.offset; });
+}
```

Comment:
Why not store the offsets and lower bounds as `isl::multi_aff`s in the first place?

Reply:
Because there is a `ScopedFootprintDim` class; changing/removing that class is extra work that does not bring us anywhere closer to the goal of this PR.